src/libutil/hash.cc

```diff
@@ -202,14 +214,16 @@ Hash::Hash(std::string_view s, HashType type)
     }
 
     else
-        throw BadHash("hash '%s' has wrong length for hash type '%s'", s, printHashType(type));
+        throw BadHash("hash '%s' has wrong length for hash type '%s'", s, printHashType(*type));
 }
 
-Hash newHashAllowEmpty(std::string hashStr, HashType ht)
+Hash newHashAllowEmpty(std::string hashStr, std::optional<HashType> ht)
 {
     if (hashStr.empty()) {
-        Hash h(ht);
-        warn("found empty hash, assuming '%s'", h.to_string(SRI, true));
+        if (!ht)
+            throw BadHash("empty hash requires explicit hash type");
```

Comment:
This is a new validation, actually. I think we want it.

CC @matthewbauer

Reply:
it's okay if hash type is unset - you just have to have sri in your string

Reply:
Yeah this is just in the empty string case, when the SRI or legacy prefix can't help.